project1/client/client.cpp

```diff
@@ -174,21 +128,49 @@ bool Client::FtpShell() {
 
     // create & serialize request message for determined command
     r = ip_->CreateReq();
-    wire_protocol::Serialize(r, &outgoing_msg_buf_);
-
+    if (!ip_->IsValid()) {
+      std::cout << "Invalid command, try again." << "\n";
+      continue;
+    }
+    if (r.has_terminate()) {
+      auto terminate_task = std::make_shared<client_tasks::TerminateTask>(hostname_, tport_, r.terminate());
+      pool.AddTask(terminate_task);
+      delete ip_;
```

Comment:
Much better to use a `unique_ptr` to store this instead of manually managing memory.

Reply:
I added this in 06164ccfeae0a8ced9aa0f3638868120c13f076d. I believe I did it correctly, but will need to be given a quick glance.

Reply:
Noted. I'll look.